File: ChatAppClient/src/github/chatapp/client/main/java/service/client/io_client/MessageHandler.java
/* Copyright (C) 2023 Ilias Koukovinis <ilias.koukovinis@gmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 * 
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */
package github.chatapp.client.main.java.service.client.io_client;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.atomic.AtomicBoolean;

import github.chatapp.client.main.java.service.client.ChatRequest;
import github.chatapp.client.main.java.service.client.ChatSession;
import github.chatapp.client.main.java.service.client.DonationHtmlPage;
import github.chatapp.common.LoadedInMemoryFile;
import github.chatapp.common.message_types.ClientCommandResultType;
import github.chatapp.common.message_types.ClientCommandType;
import github.chatapp.common.message_types.ClientMessageType;
import github.chatapp.common.message_types.ContentType;
import github.chatapp.common.message_types.Message;
import github.chatapp.common.message_types.ServerMessageType;
import github.chatapp.common.util.EnumIntConverter;
import io.netty.buffer.ByteBuf;
import io.netty.buffer.Unpooled;

/**
 * 
 * @author Ilias Koukovinis
 * 
 */
public abstract class MessageHandler implements AutoCloseable {

	private ByteBufInputStream in;
	private ByteBufOutputStream out;

	private String username;
	private int clientID;
	
	private Map<Integer, ChatSession> chatSessionIDSToChatSessions = new HashMap<>();
	private List<ChatSession> chatSessions = new ArrayList<>();
	private List<ChatRequest> chatRequests = new ArrayList<>();

	private AtomicBoolean isClientListeningToMessages = new AtomicBoolean(false);
	private Commands commands = new Commands();
	
	void setByteBufInputStream(ByteBufInputStream in) {
		this.in = in;
	}
	
	void setByteBufOutputStream(ByteBufOutputStream out) {
		this.out = out;
	}
	
	public void sendMessageToClient(String text, int chatSessionIndex) throws IOException {

		byte[] textBytes = text.getBytes();
		
		ByteBuf payload = Unpooled.buffer();
		payload.writeInt(EnumIntConverter.getEnumAsInt(ClientMessageType.CLIENT_CONTENT));
		payload.writeInt(EnumIntConverter.getEnumAsInt(ContentType.TEXT));
		payload.writeInt(chatSessionIndex);
		payload.writeInt(textBytes.length);
		payload.writeBytes(textBytes);
		
		out.write(payload);
	}
	
	public void sendFile(File file, int chatSessionIndex) throws IOException {

		byte[] fileNameBytes = file.getName().getBytes();
		byte[] fileBytes = Files.readAllBytes(Paths.get(file.getAbsolutePath()));

		ByteBuf payload = Unpooled.buffer();
		payload.writeInt(EnumIntConverter.getEnumAsInt(ClientMessageType.CLIENT_CONTENT));
		payload.writeInt(EnumIntConverter.getEnumAsInt(ContentType.FILE));
		payload.writeInt(chatSessionIndex);
		payload.writeInt(fileNameBytes.length);
		payload.writeBytes(fileNameBytes);
		payload.writeBytes(fileBytes);

		out.write(payload);
	}

	public abstract void usernameReceived(String username);
	public abstract void messageReceived(Message message, int chatSessionIndex);
	public abstract void alreadyWrittenTextReceived(ChatSession chatSession);
	public abstract void serverMessageReceived(String message);
	public abstract void fileDownloaded(LoadedInMemoryFile file);
	public abstract void donationPageReceived(DonationHtmlPage donationPage);
	public abstract void serverSourceCodeReceived(String serverSourceCodeURL);
	public abstract void clientIDReceived(int clientID);
	public abstract void chatRequestsReceived(List<ChatRequest> chatRequests);
	public abstract void chatSessionsReceived(List<ChatSession> chatSessions);
	public abstract void messageDeleted(ChatSession chatSession, int messageIDOfDeletedMessage);
	
	public class Commands {

		public void changeUsername(String newUsername) throws IOException {

			byte[] newUsernameBytes = newUsername.getBytes();

			ByteBuf payload = Unpooled.buffer();
			payload.writeInt(EnumIntConverter.getEnumAsInt(ClientMessageType.COMMAND));
			payload.writeInt(EnumIntConverter.getEnumAsInt(ClientCommandType.CHANGE_USERNAME));
			payload.writeBytes(newUsernameBytes);

			out.write(payload);
		}
		
		public void changePassword(String newPassword) throws IOException {
			
			byte[] newPasswordBytes = newPassword.getBytes();

			ByteBuf payload = Unpooled.buffer();
			payload.writeInt(EnumIntConverter.getEnumAsInt(ClientMessageType.COMMAND));
			payload.writeInt(EnumIntConverter.getEnumAsInt(ClientCommandType.CHANGE_PASSWORD));
			payload.writeBytes(newPasswordBytes);

			out.write(payload);
		}

		public void sendGetUsername() throws IOException {
			
			ByteBuf payload = Unpooled.buffer();
			payload.writeInt(EnumIntConverter.getEnumAsInt(ClientMessageType.COMMAND));
			payload.writeInt(EnumIntConverter.getEnumAsInt(ClientCommandType.GET_USERNAME));
			
			out.write(payload);
		}
		
		public void sendGetClientID() throws IOException {
			
			ByteBuf payload = Unpooled.buffer(Integer.BYTES * 2);
			payload.writeInt(EnumIntConverter.getEnumAsInt(ClientMessageType.COMMAND));
			payload.writeInt(EnumIntConverter.getEnumAsInt(ClientCommandType.GET_CLIENT_ID));
			
			out.write(payload);
		}
		
		public void sendGetWrittenText(int chatSessionIndex) throws IOException {
			
			ByteBuf payload = Unpooled.buffer();
			payload.writeInt(EnumIntConverter.getEnumAsInt(ClientMessageType.COMMAND));
			payload.writeInt(EnumIntConverter.getEnumAsInt(ClientCommandType.GET_WRITTEN_TEXT));
			payload.writeInt(chatSessionIndex);
			payload.writeInt(chatSessions.get(chatSessionIndex).getMessages().size() /* Amount of messages client already has */);
			
			out.write(payload);
		}
		
		public void sendGetChatRequests() throws IOException {
			
			ByteBuf payload = Unpooled.buffer();
			payload.writeInt(EnumIntConverter.getEnumAsInt(ClientMessageType.COMMAND));
			payload.writeInt(EnumIntConverter.getEnumAsInt(ClientCommandType.GET_CHAT_REQUESTS));

			out.write(payload);
		}
		
		public void sendGetChatSessions() throws IOException {

			ByteBuf payload = Unpooled.buffer();
			payload.writeInt(EnumIntConverter.getEnumAsInt(ClientMessageType.COMMAND));
			payload.writeInt(EnumIntConverter.getEnumAsInt(ClientCommandType.GET_CHAT_SESSIONS));

			out.write(payload);
		}
		
		public void sendGetDonationHTMLPage() throws IOException {

			ByteBuf payload = Unpooled.buffer();
			payload.writeInt(EnumIntConverter.getEnumAsInt(ClientMessageType.COMMAND));
			payload.writeInt(EnumIntConverter.getEnumAsInt(ClientCommandType.GET_DONATION_PAGE));
			
			out.write(payload);
		}
		
		public void sendGetServerSourceCodeHTMLPage() throws IOException {
			
			ByteBuf payload = Unpooled.buffer();
			payload.writeInt(EnumIntConverter.getEnumAsInt(ClientMessageType.COMMAND));
			payload.writeInt(EnumIntConverter.getEnumAsInt(ClientCommandType.GET_SERVER_SOURCE_CODE_PAGE));
			
			out.write(payload);
		}

		public void sendChatRequest(int userClientID) throws IOException {

			ByteBuf payload = Unpooled.buffer();
			payload.writeInt(EnumIntConverter.getEnumAsInt(ClientMessageType.COMMAND));
			payload.writeInt(EnumIntConverter.getEnumAsInt(ClientCommandType.SEND_CHAT_REQUEST));
			payload.writeInt(userClientID);
			
			out.write(payload);
		}
		
		public void acceptChatRequest(int userClientID) throws IOException {
			
			ByteBuf payload = Unpooled.buffer();
			payload.writeInt(EnumIntConverter.getEnumAsInt(ClientMessageType.COMMAND));
			payload.writeInt(EnumIntConverter.getEnumAsInt(ClientCommandType.ACCEPT_CHAT_REQUEST));
			payload.writeInt(userClientID);
			
			out.write(payload);
			
			sendGetChatRequests();
			sendGetChatSessions();
		}
		
		public void declineChatRequest(int userClientID) throws IOException {
			
			ByteBuf payload = Unpooled.buffer();
			payload.writeInt(EnumIntConverter.getEnumAsInt(ClientMessageType.COMMAND));
			payload.writeInt(EnumIntConverter.getEnumAsInt(ClientCommandType.DECLINE_CHAT_REQUEST));
			payload.writeInt(userClientID);
			
			out.write(payload);

			sendGetChatRequests();
		}
		
		public void deleteChatSession(int chatSessionIndex) throws IOException {
			
			ByteBuf payload = Unpooled.buffer();
			payload.writeInt(EnumIntConverter.getEnumAsInt(ClientMessageType.COMMAND));
			payload.writeInt(EnumIntConverter.getEnumAsInt(ClientCommandType.DELETE_CHAT_SESSION));
			payload.writeInt(chatSessionIndex);
			
			out.write(payload);
			
			sendGetChatSessions();
		}
		
		public void deleteMessage(int chatSessionIndex, int messageID) throws IOException {
			
			ByteBuf payload = Unpooled.buffer();
			payload.writeInt(EnumIntConverter.getEnumAsInt(ClientMessageType.COMMAND));
			payload.writeInt(EnumIntConverter.getEnumAsInt(ClientCommandType.DELETE_CHAT_MESSAGE));
			payload.writeInt(chatSessionIndex);
			payload.writeInt(messageID);
			
			out.write(payload);
		}
		
		public void downloadFile(int messageID, int chatSessionIndex) throws IOException {

			ByteBuf payload = Unpooled.buffer();
			payload.writeInt(EnumIntConverter.getEnumAsInt(ClientMessageType.COMMAND));
			payload.writeInt(EnumIntConverter.getEnumAsInt(ClientCommandType.DOWNLOAD_FILE));
			payload.writeInt(chatSessionIndex);
			payload.writeInt(messageID);
			
			out.write(payload);
		}
		
		public void logout() throws IOException {
			
			ByteBuf payload = Unpooled.buffer();
			payload.writeInt(EnumIntConverter.getEnumAsInt(ClientMessageType.COMMAND));
			payload.writeInt(EnumIntConverter.getEnumAsInt(ClientCommandType.LOGOUT));
			
			out.write(payload);
		}

	}
	
	/**
	 * reads incoming messages sent from the server
	 */
	public void startListeningToMessages() throws IOException {
		
		if (isClientListeningToMessages()) {
			throw new IllegalStateException("Client is already listening to messages!");
		}
		
		isClientListeningToMessages.set(true);

		Thread thread = new Thread("Thread-listenToMessages") {
			
			@Override
			public void run() {
				while (isClientListeningToMessages()) {
					try {
						
						ByteBuf msg = in.read();
						
						ServerMessageType msgType = EnumIntConverter.getIntAsEnum(msg.readInt(), ServerMessageType.class);

						switch (msgType) {
						case SERVER_MESSAGE_INFO -> {
							
							byte[] content = new byte[msg.readableBytes()];
							msg.readBytes(content);
							
							serverMessageReceived(new String(content));
						}
						case CLIENT_CONTENT -> {

							Message message = new Message();
							
							ContentType contentType = EnumIntConverter.getIntAsEnum(msg.readInt(), ContentType.class);
							
							byte[] text = null;
							byte[] fileNameBytes = null;
							
							switch (contentType) {
							case TEXT -> {
								
								text = new byte[msg.readInt()];
								msg.readBytes(text);
							}
							case FILE -> {
								
								fileNameBytes = new byte[msg.readInt()];
								msg.readBytes(fileNameBytes);
							}
							}
							
							byte[] usernameBytes = new byte[msg.readInt()];
							msg.readBytes(usernameBytes);
							
							String username = new String(usernameBytes);
							int clientID = msg.readInt();
							int messageID = msg.readInt();
							int chatSessionID = msg.readInt();
							
							message.setContentType(contentType);
							message.setUsername(username);
							message.setClientID(clientID);
							message.setMessageID(messageID);
							message.setChatSessionID(chatSessionID);
							message.setText(text);
							message.setFileName(fileNameBytes);
							
							ChatSession chatSession = chatSessionIDSToChatSessions.get(chatSessionID);
							
							if (chatSession.haveChatMessagesBeenCached()) {
								chatSession.getMessages().add(message);
							}
							
							messageReceived(message, chatSession.getChatSessionIndex());
						}
						case COMMAND_RESULT -> {

							ClientCommandResultType commandResult = EnumIntConverter.getIntAsEnum(msg.readInt(), ClientCommandResultType.class);

							switch (commandResult) {
							case DOWNLOAD_FILE -> {
								
								byte[] fileNameBytes = new byte[msg.readInt()];
								msg.readBytes(fileNameBytes);
								
								byte[] fileBytes = new byte[msg.readableBytes()];
								msg.readBytes(fileBytes);
								
								fileDownloaded(new LoadedInMemoryFile(new String(fileNameBytes), fileBytes));
							}
							case GET_USERNAME -> {
								
								byte[] usernameBytes = new byte[msg.readableBytes()];
								msg.readBytes(usernameBytes);

								username = new String(usernameBytes);
								
								usernameReceived(username);
							}
							case GET_CLIENT_ID -> {
								clientID = msg.readInt();
								clientIDReceived(clientID);
							}
							case GET_CHAT_SESSIONS -> {

								chatSessions.clear();
								
								int chatSessionsSize = msg.readInt();
								for (int i = 0; i < chatSessionsSize; i++) {

									int chatSessionIndex = i;
									int chatSessionID = msg.readInt();

									ChatSession chatSession = new ChatSession(chatSessionID, chatSessionIndex);

									int membersSize = msg.readInt();

									List<ChatSession.Member> members = new ArrayList<>(membersSize);

									for (int j = 0; j < membersSize; j++) {

										int clientID = msg.readInt();

										byte[] usernameBytes = new byte[msg.readInt()];
										msg.readBytes(usernameBytes);

										members.add(new ChatSession.Member(new String(usernameBytes), clientID));
									}

									chatSession.setMembers(members);

									chatSessions.add(chatSessionIndex, chatSession);
									chatSessionIDSToChatSessions.put(chatSessionID, chatSession);
								}
								
								chatSessionsReceived(chatSessions);
							}
							case GET_CHAT_REQUESTS -> {
								
								chatRequests.clear();
								
								int friendRequestsLength = msg.readInt();

								for (int i = 0; i < friendRequestsLength; i++) {
									int clientID = msg.readInt();
									chatRequests.add(new ChatRequest(clientID));
								}
								
								chatRequestsReceived(chatRequests);
							}
							case GET_WRITTEN_TEXT -> {

								ChatSession chatSession;

								{
									int chatSessionIndex = msg.readInt();
									
									chatSession = chatSessions.get(chatSessionIndex);
								}
								
								List<Message> messages = chatSession.getMessages();
								while (msg.readableBytes() > 0) {

									ContentType contentType = EnumIntConverter.getIntAsEnum(msg.readInt(), ContentType.class);
									
									int clientID = msg.readInt();
									int messageID = msg.readInt();

									String username;
									
									{
										byte[] usernameBytes = new byte[msg.readInt()];
										msg.readBytes(usernameBytes);
								
										username = new String(usernameBytes);
									}

									byte[] messageBytes = null;
									byte[] fileNameBytes = null;
									
									switch (contentType) {
									case TEXT -> {
										messageBytes = new byte[msg.readInt()];
										msg.readBytes(messageBytes);
									}
									case FILE -> {
										fileNameBytes = new byte[msg.readInt()];
										msg.readBytes(fileNameBytes);
									}
									}
									
									if (contentType != null) {
										Message message = new Message(username, clientID, messageID, chatSession.getChatSessionID(), messageBytes, fileNameBytes, contentType);
										messages.add(message);
									}
								}
								
								messages.sort(Comparator.comparing(Message::getMessageID));
								
								chatSession.setHaveChatMessagesBeenCached(true);
								alreadyWrittenTextReceived(chatSession);
							}
							case GET_DONATION_PAGE -> {

								byte[] htmlBytes = new byte[msg.readInt()];
								msg.readBytes(htmlBytes);
								
								byte[] htmlFileName = new byte[msg.readableBytes()];
								msg.readBytes(htmlFileName);

								DonationHtmlPage htmlPage = new DonationHtmlPage(new String(htmlBytes), new String(htmlFileName));

								donationPageReceived(htmlPage);
							}
							case GET_SERVER_SOURCE_CODE_PAGE -> {
								
								byte[] htmlPageURL = new byte[msg.readableBytes()];
								msg.readBytes(htmlPageURL);
								
								serverSourceCodeReceived(new String(htmlPageURL));
							}
							case DELETE_CHAT_MESSAGE -> {

								int chatSessionID = msg.readInt();
								int messageID = msg.readInt();

								messageDeleted(chatSessionIDSToChatSessions.get(chatSessionID), messageID);
							}
							}
						}
						}
					} catch (Exception e) {
						e.printStackTrace();
					}
				}
				
				isClientListeningToMessages.set(false);
			}
		};
		thread.setDaemon(true);
		thread.start();
		
		getCommands().sendGetUsername();
		getCommands().sendGetClientID();
		getCommands().sendGetChatSessions();
		getCommands().sendGetChatRequests();
	}

	public void stopListeningToMessages() {
		
		if (!isClientListeningToMessages()) {
			throw new IllegalStateException("Client isn't listening to messages to stop listening to messages!");
		}
		
		isClientListeningToMessages.set(false);
	}
	
	public boolean isClientListeningToMessages() {
		return isClientListeningToMessages.get();
	}
	
	public Commands getCommands() {
		return commands;
	}

	public Map<Integer, ChatSession> getChatSessionIDSToChatSessions() {
		return chatSessionIDSToChatSessions;
	}

	public List<ChatSession> getChatSessions() {
		return chatSessions;
	}

	public List<ChatRequest> getChatRequests() {
		return chatRequests;
	}

	public String getUsername() {
		return username;
	}

	public int getClientID() {
		return clientID;
	}

	@Override
	public void close() {
		stopListeningToMessages();
	}
}

File: ErmisClient/Mobile/ermis_client/lib/client/I/message_handler.dart
import 'dart:convert';
import 'dart:io';
import 'dart:typed_data';

import '../common/common.dart';
import 'byte_buf.dart';
import 'chat_request.dart';
import 'chat_session.dart';
import 'file_heap.dart';
import 'html_page.dart';
import 'input_stream.dart';
import 'message.dart';
import 'output_stream.dart';

class MessageHandler {
  late final ByteBufInputStream _inputStream;
  late final ByteBufOutputStream _outputStream;
  late final SecureSocket _secureSocket;

  late String _username;
  late int clientID;

  final Map<int, ChatSession> _chatSessionIDSToChatSessions = {};
  final List<ChatSession> _chatSessions = List.empty(growable: true);
	final List<ChatRequest> _chatRequests = List.empty(growable: true);

  bool _isClientListeningToMessages = false;
  
  late final Commands _commands;

  MessageHandler();

  void setByteBufInputStream(ByteBufInputStream inputStream) {
    _inputStream = inputStream;
  }

  void setByteBufOutputStream(ByteBufOutputStream outputStream) {
    _outputStream = outputStream;
    _commands = Commands(_outputStream, _chatSessions);
  }

  void setSecureSocket(SecureSocket secureSocket) {
    _secureSocket = secureSocket;
  }

  void sendMessageToClient(String text, int chatSessionIndex) {
    Uint8List textBytes = utf8.encode(text);

    ByteBuf payload = ByteBuf.smallBuffer();
    payload.writeInt(ClientMessageType.clientContent.id);
    payload.writeInt(ContentType.text.id);
    payload.writeInt(chatSessionIndex);
    payload.writeInt(textBytes.length);
    payload.writeBytes(textBytes);

    _outputStream.write(payload);
  }

  late void Function(String username) usernameFunction;
  void whenUsernameReceived(void Function(String username) runThis) {
    this.usernameFunction = runThis;
  }
  
  late void Function(Message message, int chatSessionIndex) messageFunction;
  void whenMessageReceived(void Function(Message message, int chatSessionIndex) runThis) {
    this.messageFunction = runThis;
  }

  late void Function(ChatSession chatSession) writtenTextFunction;
  void whenAlreadyWrittenTextReceived(void Function(ChatSession chatSession) runThis) {
    this.writtenTextFunction = runThis;
  }
  
  late void Function(String message) serverMessageReceivedFunction;
  void whenServerMessageReceived(void Function(String message) runThis) {
    this.serverMessageReceivedFunction = runThis;
  }
  
  late void Function(LoadedInMemoryFile file) whenFileDownloadedFunction;
  void whenFileDownloaded(void Function(LoadedInMemoryFile file) runThis) {
    this.whenFileDownloadedFunction = runThis;
  }

  late void Function(DonationHtmlPage donationPage) donationPageFunction;
  void whenDonationPageReceived(void Function(DonationHtmlPage donationPage) runThis) {
    this.donationPageFunction = runThis;
  }  

  late void Function(String serverSourceCodeURL) serverSourceCodeReceivedFuction;
  void whenServerSourceCodeReceived(void Function(String serverSourceCodeURL) runThis) {
    this.serverSourceCodeReceivedFuction = runThis;
  }

  late void Function(int clientID) clientIDFunction;
  void whenClientIDReceived(void Function(int clientID) runThis) {
    this.clientIDFunction = runThis;
  }

  late void Function(List<ChatRequest> chatRequests) chatRequestsReceived;
  void whenChatRequestsReceived(void Function(List<ChatRequest> chatRequests) runThis) {
    this.chatRequestsReceived = runThis;
  }

  late void Function(List<ChatSession> chatSessions) chatSessionsReceived;
  void whenChatSessionsReceived(void Function(List<ChatSession> chatSessions) runThis) {
    this.chatSessionsReceived = runThis;
  }

  late void Function(ChatSession chatSession, int messageIDOfDeletedMessage) messageDeleted;
  void whenMessageDeleted(void Function(ChatSession chatSession, int messageIDOfDeletedMessage) runThis) {
    this.messageDeleted = runThis;
  }

  late void Function(Uint8List icon) iconReceived;
  void whenIconReceived(void Function(Uint8List icon) runThis) {
    this.iconReceived = runThis;
  }

  Future<void> startListeningToMessages() async {
    if (isListeningToMessages) {
      throw StateError("Client is already listening to messages!");
    }

    _isClientListeningToMessages = true;

    _inputStream.stream.listen(
      (Uint8List data) {
        ByteBuf message = ByteBufInputStream.decode(data);
        handleMessage(message);
      },
      onDone: () {
        _secureSocket.destroy();
      },
      onError: (e) {
        print('Error: $e');
      },
    );

    getCommands().fetchUsername();
    getCommands().fetchClientID();
    getCommands().fetchChatSessions();
    getCommands().fetchChatRequests();
    // getCommands().fetchAccountIcon();
  }

  void handleMessage(ByteBuf msg){
      try {
        ServerMessageType msgType = ServerMessageType.fromId(msg.readInt32());

        switch (msgType) {
          case ServerMessageType.serverMessageInfo:
            Uint8List content = msg.readBytes(msg.readableBytes);
            serverMessageReceivedFunction(String.fromCharCodes(content));
            break;

          case ServerMessageType.clientContent:
            Message message = Message.empty();

            ContentType contentType = ContentType.fromId(msg.readInt32());
            int timeWritten = msg.readInt64();
            Uint8List? text;
            Uint8List? fileNameBytes;

            switch (contentType) {
              case ContentType.text:
                var textLength = msg.readInt32();
                text = msg.readBytes(textLength);
                break;
              case ContentType.file:
                var fileNameLength = msg.readInt32();
                fileNameBytes = msg.readBytes(fileNameLength);
                break;
            }

            var usernameLength = msg.readInt32();
            var usernameBytes = msg.readBytes(usernameLength);
            String username = String.fromCharCodes(usernameBytes);

            int clientID = msg.readInt32();
            int messageID = msg.readInt32();
            int chatSessionID = msg.readInt32();

            message.setContentType(contentType);
            message.setUsername(username);
            message.setClientID(clientID);
            message.setMessageID(messageID);
            message.setChatSessionID(chatSessionID);
            message.setText(text);
            message.setFileName(fileNameBytes);
            message.setTimeWritten(timeWritten);

            ChatSession chatSession = _chatSessionIDSToChatSessions[chatSessionID]!;
            if (chatSession.haveChatMessagesBeenCached) {
              chatSession.getMessages.add(message);
            }

            messageFunction(message, chatSession.chatSessionIndex);
            break;

          case ServerMessageType.commandResult:
            ClientCommandResultType commandResult =
                ClientCommandResultType.fromId(msg.readInt32());

            switch (commandResult) {
              case ClientCommandResultType.downloadFile:
                var fileNameLength = msg.readInt32();
                var fileNameBytes = msg.readBytes(fileNameLength);
                var fileBytes = msg.readBytes(msg.readableBytes);
                whenFileDownloadedFunction(LoadedInMemoryFile(String.fromCharCodes(fileNameBytes), fileBytes));
                break;

              case ClientCommandResultType.getDisplayName:
                var usernameBytes = msg.readBytes(msg.readableBytes);
                var username = String.fromCharCodes(usernameBytes);
                usernameFunction(username);
                break;

              case ClientCommandResultType.getClientId:
                clientID = msg.readInt32();
                clientIDFunction(clientID);
                break;

            case ClientCommandResultType.getChatSessions:
              _chatSessions.clear();
              int chatSessionsSize = msg.readInt32();
              for (int i = 0; i < chatSessionsSize; i++) {
                int chatSessionIndex = i;
                int chatSessionID = msg.readInt32();

                ChatSession chatSession =
                    ChatSession(chatSessionID, chatSessionIndex);
                int membersSize = msg.readInt32();
                List<Member> members = <Member>[];

                for (int j = 0; j < membersSize; j++) {
                  int memberClientID = msg.readInt32();
                  int usernameLength = msg.readInt32();
                  var username =
                      String.fromCharCodes(msg.readBytes(usernameLength));

                  members.add(Member(username, memberClientID));
                }

                chatSession.setMembers(members);
                _chatSessions.insert(chatSessionIndex, chatSession);
                _chatSessionIDSToChatSessions[chatSessionID] = chatSession;
              }
              chatSessionsReceived(_chatSessions);
              break;

              case ClientCommandResultType.getChatRequests:
                _chatRequests.clear();
                int friendRequestsLength = msg.readInt32();
                for (int i = 0; i < friendRequestsLength; i++) {
                  int clientID = msg.readInt32();
                  _chatRequests.add(ChatRequest(clientID));
                }
                chatRequestsReceived(_chatRequests);
                break;

              case ClientCommandResultType.getWrittenText:
                int chatSessionIndex = msg.readInt32();
                ChatSession chatSession = _chatSessions[chatSessionIndex];
                List<Message> messages = chatSession.getMessages;

                while (msg.readableBytes > 0) {
                  ContentType contentType =
                      ContentType.fromId(msg.readInt32());
                  int clientID = msg.readInt32();
                  int messageID = msg.readInt32();
                  String username = String.fromCharCodes(msg.readBytes(msg.readInt32()));

                  Uint8List? messageBytes;
                  Uint8List? fileNameBytes;
                  int timeWritten =
                      msg.readInt64();

                  switch (contentType) {
                    case ContentType.text:
                      messageBytes = msg.readBytes(msg.readInt32());
                      break;
                    case ContentType.file:
                      fileNameBytes = msg.readBytes(msg.readInt32());
                      break;
                  }

                  messages.add(Message(
                      username,
                      clientID,
                      messageID,
                      chatSession.chatSessionID,
                      messageBytes,
                      fileNameBytes,
                      timeWritten,
                      contentType));
                                  }
                messages.sort((a, b) => a.messageID.compareTo(b.messageID));
                chatSession.setHaveChatMessagesBeenCached(true);
                writtenTextFunction(chatSession);
                break;
              case ClientCommandResultType.deleteChatMessage:
                // TODO: Handle this case.
                break;
              case ClientCommandResultType.fetchAccountIcon:
                // TODO: Handle this case.
                break;
              case ClientCommandResultType.getDonationPage:
                // TODO: Handle this case.
                break;
              case ClientCommandResultType.getSourceCodePage:
                // TODO: Handle this case.
                break;
            }
            break;
        }
      } catch (e) {
        print('Error while processing message: $e');
      }

    // _isClientListeningToMessages = false;
    // Do something
  }

  bool get isListeningToMessages => _isClientListeningToMessages;
  Commands getCommands() => _commands;
}

class Commands {
  final ByteBufOutputStream out;
  final List<ChatSession> _chatSessions;

  Commands(this.out, this._chatSessions);

  void changeDisplayName(String newDisplayName) {
    var newUsernameBytes = utf8.encode(newDisplayName);

    ByteBuf payload = ByteBuf.smallBuffer();
    payload.writeInt(ClientMessageType.command.id);
    payload.writeInt(ClientCommandType.changeUsername.id);
    payload.writeBytes(newUsernameBytes);

    out.write(payload);
  }

  void changePassword(String newPassword) {
    var newPasswordBytes = utf8.encode(newPassword);

    ByteBuf payload = ByteBuf.smallBuffer();
    payload.writeInt(ClientMessageType.command.id);
    payload.writeInt(ClientCommandType.changePassword.id);
    payload.writeBytes(newPasswordBytes);

    out.write(payload);
  }

  void fetchUsername() {
    ByteBuf payload = ByteBuf.smallBuffer();
    payload.writeInt(ClientMessageType.command.id);
    payload.writeInt(ClientCommandType.fetchUsername.id);

    out.write(payload);
  }

  void fetchClientID() {
    ByteBuf payload = ByteBuf.smallBuffer();
    payload.writeInt(ClientMessageType.command.id);
    payload.writeInt(ClientCommandType.fetchClientId.id);

    out.write(payload);
  }

  void fetchWrittenText(int chatSessionIndex) {
    ByteBuf payload = ByteBuf.smallBuffer();
    payload.writeInt(ClientMessageType.command.id);
    payload.writeInt(ClientCommandType.fetchWrittenText.id);
    payload.writeInt(chatSessionIndex);
    payload.writeInt(_chatSessions[chatSessionIndex]
        .getMessages
        .length /* Amount of messages client already has */);

    out.write(payload);
  }

  void fetchChatRequests() {
    ByteBuf payload = ByteBuf.smallBuffer();
    payload.writeInt(ClientMessageType.command.id);
    payload.writeInt(ClientCommandType.fetchChatRequests.id);

    out.write(payload);
  }

  void fetchChatSessions() {
    ByteBuf payload = ByteBuf.smallBuffer();
    payload.writeInt(ClientMessageType.command.id);
    payload.writeInt(ClientCommandType.fetchChatSessions.id);

    out.write(payload);
  }

  void requestDonationHTMLPage() {
    ByteBuf payload = ByteBuf.smallBuffer();
    payload.writeInt(ClientMessageType.command.id);
    payload.writeInt(ClientCommandType.requestDonationPage.id);

    out.write(payload);
  }

  void requestServerSourceCodeHTMLPage() {
    ByteBuf payload = ByteBuf.smallBuffer();
    payload.writeInt(ClientMessageType.command.id);
    payload.writeInt(ClientCommandType.requestSourceCodePage.id);

    out.write(payload);
  }

  void sendChatRequest(int userClientID) {
    ByteBuf payload = ByteBuf.smallBuffer();
    payload.writeInt(ClientMessageType.command.id);
    payload.writeInt(ClientCommandType.sendChatRequest.id);
    payload.writeInt(userClientID);

    out.write(payload);
  }

  void acceptChatRequest(int userClientID) {
    ByteBuf payload = ByteBuf.smallBuffer();
    payload.writeInt(ClientMessageType.command.id);
    payload.writeInt(ClientCommandType.acceptChatRequest.id);
    payload.writeInt(userClientID);

    out.write(payload);

    fetchChatRequests();
    fetchChatSessions();
  }

  void declineChatRequest(int userClientID) {
    ByteBuf payload = ByteBuf.smallBuffer();
    payload.writeInt(ClientMessageType.command.id);
    payload.writeInt(ClientCommandType.declineChatRequest.id);
    payload.writeInt(userClientID);

    out.write(payload);

    fetchChatRequests();
  }

  void deleteChatSession(int chatSessionIndex) {
    ByteBuf payload = ByteBuf.smallBuffer();
    payload.writeInt(ClientMessageType.command.id);
    payload.writeInt(ClientCommandType.deleteChatSession.id);
    payload.writeInt(chatSessionIndex);

    out.write(payload);

    fetchChatSessions();
  }

  void deleteMessage(int chatSessionIndex, int messageID) {
    ByteBuf payload = ByteBuf.smallBuffer();
    payload.writeInt(ClientMessageType.command.id);
    payload.writeInt(ClientCommandType.deleteChatMessage.id);
    payload.writeInt(chatSessionIndex);
    payload.writeInt(messageID);

    out.write(payload);
  }

  void downloadFile(int messageID, int chatSessionIndex) {
    ByteBuf payload = ByteBuf.smallBuffer();
    payload.writeInt(ClientMessageType.command.id);
    payload.writeInt(ClientCommandType.downloadFile.id);
    payload.writeInt(chatSessionIndex);
    payload.writeInt(messageID);

    out.write(payload);
  }

  void logout() {
    ByteBuf payload = ByteBuf.smallBuffer();
    payload.writeInt(ClientMessageType.command.id);
    payload.writeInt(ClientCommandType.logout.id);

    out.write(payload);
  }

  Future<void> addAccountIcon(File accountIcon) async {
    ByteBuf payload = ByteBuf.smallBuffer();
    payload.writeInt(ClientMessageType.command.id);
    payload.writeInt(ClientCommandType.addAccountIcon.id);
    payload.writeBytes((await accountIcon.readAsBytes()));

    out.write(payload);
  }

  void fetchAccountIcon() {
    ByteBuf payload = ByteBuf.smallBuffer();
    payload.writeInt(ClientMessageType.command.id);
    payload.writeInt(ClientCommandType.fetchAccountIcon.id);
    out.write(payload);
  }
}